### PR TITLE
Bump llvm

### PR DESCRIPTION
**Pull request by chelini**
Things that need to be understood before merge:
1. in `dce.mlir`, it looks like the constant is not eliminated anymore; why?
2. MemAlloc effect for perfBench, is correct? Or does buffer deallocation bail out on operation marked with `AutomaticAllocationScope` trait?

### CODE REVIEW

**Comment by adam-smnk** [comment]
Added minor tweaks to fix GPU tests.
A few tests now fail due to upstream changes in deallocation pass. We'll see how to handle that later but it probably needs to be fixed upstream.

> 1. in dce.mlir, it looks like the constant is not eliminated anymore; why?

MLProgram has upstream bufferization now so, the original problem from #704 disappeared. And it looks that `symbol-dce` does not remove `llvm.mlir.global` ops. But it is fine to remove this check.

**Comment by adam-smnk** [comment]
> 2. MemAlloc effect for perfBench, is correct? Or does buffer deallocation bail out on operation marked with AutomaticAllocationScope trait?

Added `AutomaticAllocationScopeResource` to `MemAlloc` to capture the automatic scope dependency.
Also added memref read, write, and recursive effects to capture side effects on `iter_args`.

File: include/TPP/Dialect/Perf/PerfOps.td [hunk 1]
```diff
@@ -98,6 +98,8 @@ def Perf_StopTimerOp : Perf_Op<"stop_timer", []> {
 
 def Perf_BenchOp : Perf_Op<"bench",
     [AutomaticAllocationScope, SingleBlockImplicitTerminator<"perf::YieldOp">,
+     MemoryEffects<[MemWrite, MemRead, MemAlloc<AutomaticAllocationScopeResource>]>,
```
Comment (rengolin) [inline include/TPP/Dialect/Perf/PerfOps.td:101]: We don't allocate memory any more, why do we need this effect?

**Review by rengolin — COMMENTED** [review]
> Added minor tweaks to fix GPU tests.
> A few tests now fail due to upstream changes in deallocation pass. We'll see how to handle that later but it probably needs to be fixed upstream.

Is this related to Vulkan (so won't happen in the SYCL runner), or is it related to how we have our tests?

**Comment by adam-smnk** [comment]
> Is this related to Vulkan (so won't happen in the SYCL runner), or is it related to how we have our tests?

I run deallocation pass as a part of SPIR-V pipeline as unified memory can be easily used with the SPIR-V runtimes.
However, in these cases `gpu.launch_func` does not implement 'MemoryEffectOpInterface' which causes the new stricter deallocation pass to fail. It is a more general issue.

File: include/TPP/Dialect/Perf/PerfOps.td (hunk 1 above)
Comment (adam-smnk) [reply include/TPP/Dialect/Perf/PerfOps.td]: After digging a bit further, good point.
`RecursiveMemoryEffects` should be enough to "derive" memory effect from nested ops.

**Review by rengolin — APPROVED** [review]

**Pull request merged**